odl/tomo/operators/ray_trafo.py

```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+                    '{} != {}'.format(proj_space.shape,
+                                      geometry.partition.shape))
+            if proj_space.dtype != vol_space.dtype:
+                raise ValueError('`proj_space.dtype` not equal to '
+                                 '`vol_space.dtype`: {} != {}'
+                                 .format(proj_space.dtype, vol_space.dtype))
+
+        if vol_space.ndim != geometry.ndim:
+            raise ValueError('`vol_space.ndim` not equal to `geometry.ndim`: '
+                             '{} != {}'.format(vol_space.ndim,
+                                               geometry.ndim))
+
+        # Cache for input/output arrays of transforms
+        self.use_cache = kwargs.pop('use_cache', True)
+
+        # Check `impl`
+        impl = kwargs.pop('impl', None)
+        impl_type, self.__cached_impl = self._check_impl(impl)
+        self._impl_type = impl_type
+        self.__impl = impl.lower() \
+            if isinstance(impl, str) else impl_type.__name__
+
+        self._geometry = geometry
 
         # Reserve name for cached properties (used for efficiency reasons)
         self._adjoint = None
-        self._astra_wrapper = None
 
         # Extra kwargs that can be reused for adjoint etc. These must
         # be retrieved with `get` instead of `pop` above.
         self._extra_kwargs = kwargs
 
         # Finally, initialize the Operator structure
-        if variant == 'forward':
-            super(RayTransformBase, self).__init__(
-                domain=reco_space, range=proj_space, linear=True)
-        elif variant == 'backward':
-            super(RayTransformBase, self).__init__(
-                domain=proj_space, range=reco_space, linear=True)
-
-    @property
-    def impl(self):
-        """Implementation back-end for the evaluation of this operator."""
-        return self.__impl
-
-    @property
-    def geometry(self):
-        """Geometry of this operator."""
-        return self.__geometry
-
-    def _call(self, x, out=None):
-        """Return ``self(x[, out])``."""
-        if self.domain.is_real:
-            return self._call_real(x, out, **self._extra_kwargs)
-
-        elif self.domain.is_complex:
-            result_parts = [
-                self._call_real(
-                    x.real, getattr(out, 'real', None), **self._extra_kwargs
-                ),
-                self._call_real(
-                    x.imag, getattr(out, 'imag', None), **self._extra_kwargs
-                ),
-            ]
-
-            if out is None:
-                out = self.range.element()
-                out.real = result_parts[0]
-                out.imag = result_parts[1]
-
-            return out
-
-        else:
-            raise RuntimeError('bad domain {!r}'.format(self.domain))
+        super(RayTransform, self).__init__(domain=vol_space,
+                                           range=proj_space, linear=True)
 
+    @staticmethod
+    def _check_impl(impl):
```

Comment:
This guy does a lot more than mere checking. Maybe `_initialize_impl`?

Reply:
Sounds good.